flocker/node/agents/pd.py

```diff
@@ -0,0 +1,371 @@
+
+"""
+A PD implementation of the ``IBlockDeviceAPI``.
+"""
+
+import requests
+
+from bitmath import GiB, Byte
+from googleapiclient import discovery
+from googleapiclient.errors import HttpError
+from oauth2client.gce import AppAssertionCredentials
+from socket import gethostname
+from twisted.python.filepath import FilePath
+from uuid import UUID
+from zope.interface import implementer
+
+from .blockdevice import (
+    IBlockDeviceAPI, BlockDeviceVolume, AlreadyAttachedVolume, UnknownVolume,
+    UnattachedVolume
+)
+from ...common import poll_until
+
+# GCE instances have a metadata server that can be queried for information
+# about the instance the code is being run on.
+_METADATA_SERVER = u'http://169.254.169.254/computeMetadata/v1/'
+_METADATA_HEADERS = {u'Metadata-Flavor': u'Google'}
+
+
+def _get_metadata_path(path):
+    """
+    Requests a metadata path from the metadata server available within GCE.
+
+    The metadata server is a good way to query information about the currently
+    running instance and project it is in. It is also the mechanism used to
+    inject ssh public keys and service account session tokens into the VM.
+
+    :param unicode path: The path on the metadata server to query.
+
+    :returns unicode: The resulting value from the metadata server.
+    """
+    timeout_sec = 3
+    r = requests.get(_METADATA_SERVER + path,
+                     headers=_METADATA_HEADERS,
+                     timeout=timeout_sec)
+    return r.text
+
+
+# The prefix added to dataset_ids to turn them into blockdevice_ids.
+_PREFIX = 'flocker-v1-'
+
+
+def _blockdevice_id_to_dataset_id(blockdevice_id):
+    """
+    Computes a dataset_id from a blockdevice_id.
+
+    :param unicode blockdevice_id: The blockdevice_id to get the dataset_id
+        for.
+
+    :returns UUID: The corresponding dataset_id.
+    """
+    return UUID(blockdevice_id[len(_PREFIX):])
+
+
+def _dataset_id_to_blockdevice_id(dataset_id):
+    """
+    Computes a blockdevice_id from a dataset_id.
+
+    :param UUID dataset_id: The dataset_id to get the blockdevice_id for.
+
+    :returns unicode: The corresponding blockdevice_id.
+    """
+    return _PREFIX + unicode(dataset_id)
+
+
+def _extract_attached_to(disk):
+    """
+    Given a GCE disk resource, determines the unicode name of the machine that
+    it is attached to.
+
+    :param dict disk: A GCE disk resource as returned from the API.
+
+    :returns: The `unicode` name of the instance the disk is attached to or
+        `None` if it is not attached to anything.
+    """
+    # TODO(mewert): determine how this works with a disk being attached to
+    # multiple machines, update comment above.
+    users = disk.get('users', [])
+    if not users:
+        return None
+    return unicode(users[0].split('/')[-1])
+
+
+@implementer(IBlockDeviceAPI)
+class PDBlockDeviceAPI(object):
+    """
+    A PD implementation of ``IBlockDeviceAPI`` which creates block devices in a
```

Comment:
Here is the meat of the design. The only think I am expecting in design review is for you to read through this and glance at my other comments on the PR. If you want to dive into the other implementation details you can, but that will already happen in code review.